### ISSUE TITLE
Question: MIME `Content-Type` Type with grammar files of *.pegjs file extension

### ISSUE THREAD

**Issue by Martii** [pegjs/pegjs#358]
Recently we've started adding support for this parser generator into our project and I'd like to say it is a godsend in making things a little more readable in our code. A Thank You goes out to your project maintainers from our maintainers. :)

Onto the question.

I've noticed that this project [has picked the file extension of *.pegjs for the grammar files...](https://github.com/pegjs/pegjs/blob/c13cc882626a245a7a5a97af036f95e40e50a2c7/examples/javascript.pegjs)

*  We were wondering if there is a standardized [MIME Type](https://developer.mozilla.org/docs/Web/Security/Securing_your_site/Configuring_server_MIME_types) that has been decided on for these grammar files?

Currently [we are using](https://github.com/OpenUserJS/OpenUserJS.org/blob/d1c4c12f9b517c795bcc6ce61c378b445c3cb1fa/app.js#L103) `text/javascript` however I have noticed that there is some uniqueness in the grammar language that doesn't appear to conform fully to JavaScript specifications... so this may not be the right type but is somewhat close.

The reason why I ask is when we currently host a set of grammar files with an extension of `.pegjs` on the server. When it is typed into the address bar of the browser... the browser t typically attempts to download the file as a BIN file *(default action for unknown Content-Type is binary in this browser)*. So rather than **risk a security warning from some other site service** we coerce its value to one that is clear *text* *(so we can read it and possibly minify it at some future date)* and the nearest match for what the syntax is.

This MIME type *(a.k.a. `Content-Type` request header)* can also be used in web and external editors to determine Syntax Highlighting. Since peg.js grammar files don't appear to registered with W3C/IANA/IETF I presume that the `x-` needs to be done. *(see below possible suggestions if I may)*.

So basically the question reiterated is.... **What should we be using according to this projects official specification for grammar files with MIME Type?**

Possible suggested naming for MIME/Content Type:
* `text/x-pegjs-grammar` *(We haven't tried to see what the server/browser does with this natively yet but will notate when tested)*
* `text/x-javascript-pegjs-grammar` *(not entirely sure about this one but may work with most programs out there that prefer MIME and start matching at the beginning and choose JavaScript as the nearest match for Syntax Highlighting or other operation.)*

**Comment by dmajda** [comment]
I’m glad that PEG.js serves you well :-)

As for the MIME type, PEG.js grammars are not JavaScript, so `text/javascript` is technically incorrect. I’d suggest `text/x-pegjs`.

For further questions please use the [PEG.js Google Group](https://groups.google.com/forum/#!forum/pegjs). Issues are intended for bug reports and feature requests. Thanks.

**Comment by Martii** [comment]
> I’d suggest `text/x-pegjs.`

Great... thank you so much.

> Issues are intended for bug reports and feature requests. Thanks.

Awesome... feature request is to have this placed in the [README.md](https://github.com/pegjs/pegjs/blob/master/README.md) or elsewhere for the docs then.

**Comment by dmajda** [comment]
I just found that according to [RFC 6838, Section 3.4](https://tools.ietf.org/html/rfc6838#section-3.4) and [RFC 6648](https://tools.ietf.org/html/rfc6648), using the `x-` prefix for MIME types is deprecated. This makes the `text/x-pegjs` recommendation obsolete.

Now, reading through [RFC 6838, Section 3](https://tools.ietf.org/html/rfc6838#section-3) it doesn’t look like any registration tree is suitable for PEG.js grammars. The standards tree, vendor tree, and personal tree all require registration with IANA, which I’m not going to go through. And the unregistered `x.` tree is “for types intended exclusively for use in private, local environments”. So, at this point, I can’t really recommend any MIME type for PEG.js grammars beside `text/plain`.

**Comment by Martii** [comment]
Well that is something I seem to have missed too... a lot of developers aren't going to be happy with this... although finding the adoption date for the RFC *(request for change)* might be prudent... I don't exactly see that other than the RFC date itself.

3.1 in the below quote uses an abstract "magic" word of "SHOULD" as well... which doesn't mean absolute... but that's a grammar thing.

> 3.  Recommendations for Creators of New Parameters
>
>   Creators of new parameters to be used in the context of application
   protocols:
>
>   1.  SHOULD assume that all parameters they create might become
       standardized, public, commonly deployed, or usable across
       multiple implementations.
>
>   2.  SHOULD employ meaningful parameter names that they have reason to
       believe are currently unused.
>
>   3.  SHOULD NOT prefix their parameter names with "X-" or similar
       constructs.
>
>   Note: If the relevant parameter name space has conventions about
   associating parameter names with those who create them, a parameter
   name could incorporate the organization's name or primary domain name
   (see Appendix B for examples).



> So, at this point, I can’t really recommend any MIME type for PEG.js grammars beside `text/plain`.

Well 3.1 "should" apply here. ;) Have something in mind that "should be created"?

Thanks for the followup. :)

**Comment by dmajda** [comment]
> although finding the adoption date for the RFC *(request for change)* might be prudent... I don't exactly see that other than the RFC date itself.

As far as I can tell, a RFC starts to apply immediately after it is published.

> Well 3.1 "should" apply here. ;)

Well, the most authoritative text is [RFC 6838, Section 3.4](https://tools.ietf.org/html/rfc6838#section-3.4), which says quite clearly, that `x-` prefixed types do not belong to the unregistered tree. Together with the rest of Section 3 this implies they fall into the standards tree, which requires registration.

**Comment by Martii** [comment]
> which says quite clearly, that `x-` prefixed types do not belong to the unregistered tree.

Semantics aside, because my comprehension seems a bit different than what you are saying... using `text/plain` may inhibit development of, at the very least, syntax highlighting routines, in which I have been assisting with development in other projects throughout the decades, and other MIME type applications for it. There is always the fallback to basename.extension detection but that methodology is also "strongly discouraged".  There are some projects I know for example that utilizes the x. tree and those additons are more recent than 1995.

I think the adoption rate for what "might" be used is determinate on how many people utilize a particular naming schema... and someone besides this project may register it for you at a later date... hence the magic word of "might" in the RFC. So if you were to "decide" on  a variant or exact of my original suggestion [above](https://github.com/pegjs/pegjs/issues/358#issue-96142437) we would definitely be a party to moving something forward with some "standardization". I understand that you may be reluctant to register... I'm not sure I would want to either but I think we're going to stick with it across the projects that I know use it... it is highly unlikely that a collision would occur especially with the grandfathering. My experienced guess is that the complaint in the RFC is because they are lazily coding with matches in whatever systems they use.

**Comment by Martii** [comment]
Here's an example of a newer x. tree addition....

```
x-matroska
```

that's part of the ogg system specific to audio with [matroska](https://www.matroska.org/technical/guides/faq/index.html). I am quite sure this post-dates the RFC. There appears to be hundreds in the *nix tree file associations as well... some older some newer.

**Comment by dmajda** [comment]
As I said, I don’t want to go through the registration procedure, and that includes also helping somebody else going through it. I’d like to to spend my time on other, more important things.

In any case, RFCs are just recommendations and you are free to ignore them. If you don’t care about compliance, you can use my original suggested type or e.g. `application/pegjs` freely. It’s just that I don’t want to encourage this officially (and put it into documentation). For private use, it’s completely legal to use e.g. `application/x.pegjs`. (I think `application` top-level type is better than `text` after reading through text in [RFC 6838, Section 4.2](https://tools.ietf.org/html/rfc6838#section-4.2).)

And yes, the media type registration procedure is inflexible and I think it undermines usability of media types. I’d much rather see something less rigid in place.
